LMMS/lmms
# MIDI Program Change message support in VeSTige, SF2 Player and GIG Player

This is a port of original pull request (#5158) which was implemented on the old code base. The original pull request is to be closed as a stale review.

**Adds the native support for the program change messages, implementing the support for the following MIDI events:
- Program Change
- Controller Change for CC IDs 0 and 32, which are Bank Select MSB and LSB, respectively.**

VeSTige, GIG Player and SF2 Player have been adjusted to make use of the new functionality, as they work natively on programs (patches).

**Description of new feature**

The new control can be found on MIDI tab, labelled _ENABLE PRESET SELECTION_. When the toggle button is on, MIDI Program Change messages are intercepted by LMMS and utilized to change patches.

The combo box inside the section allows to choose the way Bank Select messages are handled:
- _Ignore Bank Select_ - Bank Select messages are ignored. Only first 128 patches can be selected directly from the MIDI controller. This setting is useful when the MIDI keyboard or whatever device we use does not support preset banks, utilizing CCs 0 and 32 for different purposes.
- _Use Bank Select MSB only_ - Bank Select MSB controller is intercepted, its value is treated as the bank number.
- _Use Bank Select MSB and LSB_ - Both Bank Select controllers are intercepted and the value is combined to obtain the bank number between 0 and 16383.

For VeSTige, the target preset ID is determined from the bank number and the program number. Presets 0 to 127 belong to bank 0, presets 128..255 fall into bank 1, presets 256..383 belong to bank 2, and so on. GIG and SF2 players support banking natively and have separate controls for them.

Reception of the Program Change and/or Bank Select messages will trigger the update of the GUI, so the preset name and the associated values are visible immediately.

**Validation of changes**
_Preparation_
- Connect any MIDI controller capable of sending subject messages (e. g. Alesis QX49)
- Load the plugin of your choice:
  - For VeSTige: load any VST plugin having at least a few presets.
  - For GIG or SF2: load the soundfont/sample pack
- Enable the MIDI input for the plugin. If you use ALSA Sequencer as MIDI interface, pick appropriate MIDI device from the list.
- Enable the preset selection.
- Switch back to the Plugin tab to see the effects

_Setting up MIDI keyboard - example for Alesis QX49_
- Enter ADVANCED mode, hit P1 pad and configure the following settings:
  - CTRL ASSIGN: 153 (cycle through values in a range)
  - DATA1 (CC ID): 0 (Bank Select MSB)
  - DATA2 (Starting value): 0
  - DATA3 (Ending value): 7
- Still in ADVANCED mode, move S1 slider and configure the following settings:
  - CTRL ASSIGN: 32 (Bank Select LSB)
  - DATA2 (Value when the slider is in the downmost position): 0
  - DATA3 (Value when the slider is in the upmost position): 127
- Exit ADVANCED mode, press PROGRAM button to be able to cycle through programs using -/+ keys
 
_Validating Program Change_
- On the MIDI tab disable the support for Bank Select messages.
- Trigger a Program Change message on the MIDI keyboard selecting patch _k_ and observe the behavior:
  - For SF2 and GIG players, the selected patch number _k_ should be visible in the GUI.
  - For VeSTige, as it counts from 1, the patch number _k+1_ should be visible in the GUI
  - For SF2 and GIG players, the changes should not affect the bank number chosen manually
  - For VeSTige, only first 128 patches should be available.
 
_Validating Bank Select - 7-bit selection_
- On the MIDI tab enable the support for Bank Select messages - MSB only
- Change the bank number to _n_ and observe the behavior:
  - For SF2 and GIG players, the selected bank number _n_ should be visible in the GUI, the patch number should be still equal to _k_.
  - For VeSTige, the active preset's number should be _128n+k+1_. It is recommended to use a plugin that exposes more than 128 presets.

_Validating Bank Select - 14-bit selection_
- On the MIDI tab enable the support for Bank Select messages - MSB and LSB
- Change the bank number MSB (_n_) and LSB (_m_) and observe the behavior:
  - For SF2 and GIG players, the selected bank number _128n+m_ should be visible in the GUI, the patch number should be still equal to _k_.
  - For VeSTige, the active preset's number should be _16384n+128m+k+1_. This might be tricky to test, as it would be hard to find a plugin equipped with over 16000 factory patches :)
 
_Validating save & load_
- Set any state of the Program Selection and Bank Selection controls
- Save the project
- Open a new project
- Reopen the previously saved project
- Open the virtual instrument you have previously configured
  - the state of switches should be the same as when the project was saved
